### PR TITLE
implement gatsby-plugin-jss

### PR DESCRIPTION
**Pull request by wizardzloy**
this also adds a new lifecycle callback `onClientRender` to Browser API.
It is called when the initial render of a Gatsby App is done on the client.

### CODE REVIEW

**Comment by KyleAMathews** [comment]
This looks awesome! Thanks!

Any reason not to use https://www.gatsbyjs.org/docs/browser-apis/#onClientEntry?

**Comment by wizardzloy** [comment]
@KyleAMathews, when `onClientEntry` is called the `<style>` tag generated by JSS on the client is not yet in the DOM, so if at this moment we remove the one generated on the server there will be no styles applied to the html until `ReactDOM.render` is executed and JSS inserted the "client-side" styles to the DOM. This may cause the page to flash from styled view to unstyled and back to styled  again.

**Comment by KyleAMathews** [comment]
So from this discussion it looks like they've solved the SSR/client hydration issues for the next version of JSS https://github.com/cssinjs/react-jss/pull/107

I'd rather not add an special-case API to work around a bug in another library... could we just add JSS support with the client styles hydration not quite working knowing this will get fixed soon on the next release?

**Comment by KyleAMathews** [comment]
@wizardzloy Ok with taking out the new API? Would love to get this in soon!

**Comment by KyleAMathews** [comment]
Also could you add a very simple example site similar to the example sites for other css plugins.

**Comment by wizardzloy** [comment]
@kof could you please confirm that with `react-jss` v7.0.0 we don't need to remove the SSR-generated `<style>` tag after rendering on the client?

@KyleAMathews 
> Also could you add a very simple example site similar to the example sites for other css plugins.

Could you please provide a link to one of those for a reference?

**Comment by kof** [comment]
Removing  critical css style tag is still required after application was mounted. Just added a note to the readme of react-jss. It was previously documented in the core only. 

We never perceived this as a bug and had no plans to change it for now for basically 2 reasons:
1. it is easy to remove a style node
2. It gives you some freedom in how you would like to implement ssr styles and critical styles.

**Comment by KyleAMathews** [comment]
@kof I don't understand why removing the SSR styles is necessary. Are styles generated differently during SSR?  Why would it matter if they're there or not?

**Comment by kof** [comment]
> @kof I don't understand why removing the SSR styles is necessary. Are styles generated differently during SSR? Why would it matter if they're there or not?

JSS has http://cssinjs.org/json-api?v=v8.1.0#function-values which is mutative. On the server, one specific state is used, on the client, state might change over time, but the same CSS rule is used which has the same selector, but a different property.

**Comment by KyleAMathews** [comment]
But if the initial props used are the same on both server and client we're fine?

**Comment by kof** [comment]
yes, as long as props don't change at runtime, also not in the future.

**Comment by wizardzloy** [comment]
@KyleAMathews I really like the idea of keeping the API surface as small as possible but doesn't `onClientRender` look quite logical to allow the plugins to hook to? We build Gatsby on top of React and this hook maps perfectly to the React's API do you see any particular issue with having it onboard?

**Comment by kof** [comment]
Mb it helps if you show more useful use cases. I understand the hesitation for adding a new hook function which needs to be maintained when there is only one particular use case which sounds like an optional. Mb its good to see a general usefulness using some examples.

File: packages/gatsby-plugin-jss/package.json [hunk 1]
```diff
@@ -1,17 +1,24 @@
 {
   "name": "gatsby-plugin-jss",
-  "version": "1.0.1",
-  "description": "Stub description for gatsby-plugin-jss",
+  "version": "2.0.0",
```
Comment (KyleAMathews) [inline packages/gatsby-plugin-jss/package.json]: Change the version to 1.0.0

**Comment by KyleAMathews** [comment]
>yes, as long as props don't change at runtime, also not in the future.

Ok, so to be safe they'd definitely need removed.

This is a low-risk API to add and there's perhaps other use cases and I'd definitely like to support JSS. Let's change the name of the API though to `onInitialClientRender` as the current name makes it seem like it'll be called on every render.

**Comment by wizardzloy** [comment]
@KyleAMathews Thanks for the review and feedback, I changed the name of the hook to `onInitialClientRender`, added `using-jss` website to the `examples`, updated the version of `react-jss` to the latest stable version - `7.0.2`

**Comment by KyleAMathews** [comment]
Awesome! Really happy to have a JSS plugin added!

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit ac3d08f3f8451c65e4ec7acb4534f32009bd4047

https://app.netlify.com/sites/using-glamor/deploys/59778bca424ef2018a48eabd

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit ac3d08f3f8451c65e4ec7acb4534f32009bd4047

https://app.netlify.com/sites/using-contentful/deploys/59778bcb424ef2018a48eac3

**Comment by jlengstorf** [comment]
Hiya @wzrdzl! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs

**Pull request merged**